File: tools/genlayer_interact.mjs
import { createClient, createAccount } from 'genlayer-js';
import { studionet } from 'genlayer-js/chains';
import { TransactionStatus } from 'genlayer-js/types';

const RPC_URL = process.env.GENLAYER_RPC_URL;
const CONTRACT = process.env.CONTRACT_ADDRESS;
const PRIVATE_KEY = process.env.GENLAYER_PRIVATE_KEY;

if (!RPC_URL) throw new Error('GENLAYER_RPC_URL not set');
if (!CONTRACT) throw new Error('CONTRACT_ADDRESS not set');

const clientConfig = {
  chain: { ...studionet, rpcUrls: { default: { http: [RPC_URL] } } },
};

async function main() {
  const [cmd, ...args] = process.argv.slice(2);
  if (!cmd) throw new Error('Command required');

  const getArg = (flag, fallback = undefined) => {
    const i = args.indexOf(flag);
    if (i !== -1 && i + 1 < args.length) return args[i + 1];
    return fallback;
  };

  const toJson = (value) =>
    JSON.stringify(
      value,
      (_k, v) => (typeof v === 'bigint' ? v.toString() : v),
      2,
    );

  const parsePrivateKey = (pk) => {
    if (!pk) return pk;
    if (typeof pk !== 'string') return pk;
    const hex = pk.startsWith('0x') ? pk.slice(2) : pk;
    if (!/^[0-9a-fA-F]{64}$/.test(hex)) {
      throw new Error('GENLAYER_PRIVATE_KEY must be 32-byte hex (64 chars)');
    }
    return `0x${hex}`;
  };

  if (cmd === 'create') {
    if (!PRIVATE_KEY) throw new Error('GENLAYER_PRIVATE_KEY not set');
    const account = createAccount(parsePrivateKey(PRIVATE_KEY));
    const client = createClient({ ...clientConfig, account });
    const prediction = getArg('--prediction');
    const deadline = getArg('--deadline');
    const category = getArg('--category', '');
    const criteria = getArg('--criteria');
    const stake = BigInt(getArg('--stake', '0'));
    if (!prediction || !deadline || !criteria) throw new Error('Missing required args');

    const hash = await client.writeContract({
      address: CONTRACT,
      functionName: 'create_wager',
      args: [prediction, Number(stake), deadline, category, criteria],
      value: stake,
    });
    const receipt = await client.waitForTransactionReceipt({
      hash,
      status: TransactionStatus.ACCEPTED,
      retries: 50,
      interval: 4000,
    });
    console.log(toJson({ hash, receipt }));
    return;
  }

  if (cmd === 'accept') {
    if (!PRIVATE_KEY) throw new Error('GENLAYER_PRIVATE_KEY not set');
    const account = createAccount(parsePrivateKey(PRIVATE_KEY));
    const client = createClient({ ...clientConfig, account });
    const wager = getArg('--wager');
    const stake = BigInt(getArg('--stake', '0'));
    if (!wager) throw new Error('Missing --wager');
    const hash = await client.writeContract({
      address: CONTRACT,
      functionName: 'accept_wager',
      args: [wager],
      value: stake,
    });
    const receipt = await client.waitForTransactionReceipt({
      hash,
      status: TransactionStatus.ACCEPTED,
      retries: 50,
      interval: 4000,
    });
    console.log(toJson({ hash, receipt }));
    return;
  }

  if (cmd === 'verify') {
    if (!PRIVATE_KEY) throw new Error('GENLAYER_PRIVATE_KEY not set');
    const account = createAccount(parsePrivateKey(PRIVATE_KEY));
    const client = createClient({ ...clientConfig, account });
    const wager = getArg('--wager');
    const evidence = getArg('--evidence-url', '');
    if (!wager) throw new Error('Missing --wager');
    const hash = await client.writeContract({
      address: CONTRACT,
      functionName: 'submit_verification',
      args: [wager, evidence],
    });
    const receipt = await client.waitForTransactionReceipt({
      hash,
      status: TransactionStatus.ACCEPTED,
      retries: 50,
      interval: 4000,
    });
    console.log(toJson({ hash, receipt }));
    return;
  }

  if (cmd === 'appeal') {
    if (!PRIVATE_KEY) throw new Error('GENLAYER_PRIVATE_KEY not set');
    const account = createAccount(parsePrivateKey(PRIVATE_KEY));
    const client = createClient({ ...clientConfig, account });
    const wager = getArg('--wager');
    const reason = getArg('--reason');
    const evidence = getArg('--evidence-url', '');
    if (!wager || !reason) throw new Error('Missing --wager or --reason');
    const hash = await client.writeContract({
      address: CONTRACT,
      functionName: 'submit_appeal',
      args: [wager, reason, evidence],
    });
    const receipt = await client.waitForTransactionReceipt({
      hash,
      status: TransactionStatus.ACCEPTED,
      retries: 50,
      interval: 4000,
    });
    console.log(toJson({ hash, receipt }));
    return;
  }

  if (cmd === 'resolve') {
    if (!PRIVATE_KEY) throw new Error('GENLAYER_PRIVATE_KEY not set');
    const account = createAccount(parsePrivateKey(PRIVATE_KEY));
    const client = createClient({ ...clientConfig, account });
    const wager = getArg('--wager');
    if (!wager) throw new Error('Missing --wager');
    const hash = await client.writeContract({
      address: CONTRACT,
      functionName: 'resolve_wager',
      args: [wager],
    });
    const receipt = await client.waitForTransactionReceipt({
      hash,
      status: TransactionStatus.ACCEPTED,
      retries: 50,
      interval: 4000,
    });
    console.log(toJson({ hash, receipt }));
    return;
  }

  if (cmd === 'get') {
    const client = createClient({ ...clientConfig });
    const wager = getArg('--wager');
    if (!wager) throw new Error('Missing --wager');
    const result = await client.readContract({
      address: CONTRACT,
      functionName: 'get_wager',
      args: [wager],
      jsonSafeReturn: true,
    });
    console.log(toJson({ result }));
    return;
  }

  if (cmd === 'getstatus') {
    const client = createClient({ ...clientConfig });
    const wager = getArg('--wager');
    if (!wager) throw new Error('Missing --wager');
    const result = await client.readContract({
      address: CONTRACT,
      functionName: 'get_status',
      args: [wager],
      jsonSafeReturn: true,
    });
    console.log(toJson({ result }));
    return;
  }

  if (cmd === 'getjson') {
    const client = createClient({ ...clientConfig });
    const wager = getArg('--wager');
    if (!wager) throw new Error('Missing --wager');
    const result = await client.readContract({
      address: CONTRACT,
      functionName: 'get_wager_json',
      args: [wager],
    });
    console.log(toJson({ result }));
    return;
  }

  if (cmd === 'getstatusjson') {
    const client = createClient({ ...clientConfig });
    const wager = getArg('--wager');
    if (!wager) throw new Error('Missing --wager');
    const result = await client.readContract({
      address: CONTRACT,
      functionName: 'get_status_json',
      args: [wager],
    });
    console.log(toJson({ result }));
    return;
  }

  if (cmd === 'getlast') {
    const client = createClient({ ...clientConfig });
    const result = await client.readContract({
      address: CONTRACT,
      functionName: 'get_last_wager_id',
      args: [],
    });
    console.log(toJson({ result }));
    return;
  }

  throw new Error(`Unknown command: ${cmd}`);
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});

File: tools/genlayer_interact.py
import json
import os
import sys
from typing import Any, Dict, List, Optional

import requests


GENLAYER_RPC_URL = os.getenv("GENLAYER_RPC_URL")
GENLAYER_RPC_METHOD = os.getenv("GENLAYER_RPC_METHOD", "gen_call")
GENLAYER_API_KEY = os.getenv("GENLAYER_API_KEY")
CONTRACT_ADDRESS = os.getenv("CONTRACT_ADDRESS")


def rpc_call(method: str, params: Dict[str, Any]) -> Dict[str, Any]:
    if not GENLAYER_RPC_URL:
        raise RuntimeError("GENLAYER_RPC_URL not set")
    payload = {"jsonrpc": "2.0", "id": 1, "method": method, "params": params}
    headers = {"Content-Type": "application/json"}
    if GENLAYER_API_KEY:
        headers["Authorization"] = f"Bearer {GENLAYER_API_KEY}"
    r = requests.post(GENLAYER_RPC_URL, json=payload, headers=headers, timeout=20)
    r.raise_for_status()
    return r.json()


def call_contract(contract: str, fn: str, args: List[Any], value: Optional[int] = None) -> Dict[str, Any]:
    def build_params(
        contract_key: str,
        method_key: str,
        args_key: str,
    ) -> Dict[str, Any]:
        p: Dict[str, Any] = {contract_key: contract, method_key: fn, args_key: args}
        if value is not None:
            p["value"] = value
        return p

    contract_keys = ["contract", "address", "contract_address"]
    method_keys = ["method", "function", "fn", "method_name", "entrypoint"]
    args_keys = ["args", "arguments", "params", "inputs"]

    # Try the most common first, then fall back across variants.
    attempts = [
        ("contract", "method", "args"),
    ]
    for ck in contract_keys:
        for mk in method_keys:
            for ak in args_keys:
                combo = (ck, mk, ak)
                if combo not in attempts:
                    attempts.append(combo)

    last_res: Dict[str, Any] = {}
    for ck, mk, ak in attempts:
        params = build_params(ck, mk, ak)
        res = rpc_call(GENLAYER_RPC_METHOD, params)
        last_res = res
        if "error" not in res:
            return res
        err = res.get("error", {})
        msg = (err.get("message") or "").lower()
        # If error is not about unexpected parameters, stop early.
        if "unexpected parameter" not in msg:
            return res

    # Fallback: try positional params (some RPCs expect arrays)
    positional_variants: List[List[Any]] = [
        [contract, fn, args],
        [contract, fn, args, value] if value is not None else [contract, fn, args],
    ]
    for params_list in positional_variants:
        payload = {"jsonrpc": "2.0", "id": 1, "method": GENLAYER_RPC_METHOD, "params": params_list}
        headers = {"Content-Type": "application/json"}
        if GENLAYER_API_KEY:
            headers["Authorization"] = f"Bearer {GENLAYER_API_KEY}"
        r = requests.post(GENLAYER_RPC_URL, json=payload, headers=headers, timeout=20)
        r.raise_for_status()
        res = r.json()
        last_res = res
        if "error" not in res:
            return res

    return last_res


def main(argv: List[str]) -> None:
    if len(argv) < 2:
        raise SystemExit(
            "Usage:\n"
            "  python tools/genlayer_interact.py create --prediction ... --stake 100 --deadline ...\n"
            "  python tools/genlayer_interact.py accept --wager ... --stake 100 --stance agree|disagree\n"
            "  python tools/genlayer_interact.py verify --wager ... --evidence-url ...\n"
            "  python tools/genlayer_interact.py appeal --wager ... --reason ... --evidence-url ...\n"
            "  python tools/genlayer_interact.py resolve --wager ...\n"
            "  python tools/genlayer_interact.py get --wager ...\n"
        )

    if not CONTRACT_ADDRESS:
        raise SystemExit("CONTRACT_ADDRESS not set")

    cmd = argv[1]
    args = argv[2:]

    def get_arg(flag: str, default: Optional[str] = None) -> Optional[str]:
        if flag in args:
            i = args.index(flag)
            if i + 1 < len(args):
                return args[i + 1]
        return default

    if cmd == "create":
        prediction = get_arg("--prediction")
        deadline = get_arg("--deadline")
        category = get_arg("--category", "")
        criteria = get_arg("--criteria")
        stake = int(get_arg("--stake", "0") or "0")
        if not prediction or not deadline or not criteria:
            raise SystemExit("Missing --prediction, --deadline, or --criteria")
        res = call_contract(
            CONTRACT_ADDRESS,
            "create_wager",
            [prediction, stake, deadline, category, criteria],
            value=stake,
        )
    elif cmd == "accept":
        wager_id = get_arg("--wager")
        stance = get_arg("--stance", "disagree")
        stake = int(get_arg("--stake", "0") or "0")
        if not wager_id:
            raise SystemExit("Missing --wager")
        res = call_contract(CONTRACT_ADDRESS, "accept_wager", [wager_id, stance], value=stake)
    elif cmd == "verify":
        wager_id = get_arg("--wager")
        evidence_url = get_arg("--evidence-url", "")
        if not wager_id:
            raise SystemExit("Missing --wager")
        res = call_contract(CONTRACT_ADDRESS, "submit_verification", [wager_id, evidence_url])
    elif cmd == "appeal":
        wager_id = get_arg("--wager")
        reason = get_arg("--reason")
        evidence_url = get_arg("--evidence-url", "")
        if not wager_id or not reason:
            raise SystemExit("Missing --wager or --reason")
        res = call_contract(CONTRACT_ADDRESS, "submit_appeal", [wager_id, reason, evidence_url])
    elif cmd == "resolve":
        wager_id = get_arg("--wager")
        if not wager_id:
            raise SystemExit("Missing --wager")
        res = call_contract(CONTRACT_ADDRESS, "resolve_wager", [wager_id])
    elif cmd == "get":
        wager_id = get_arg("--wager")
        if not wager_id:
            raise SystemExit("Missing --wager")
        res = call_contract(CONTRACT_ADDRESS, "get_wager", [wager_id])
    elif cmd == "getstatus":
        wager_id = get_arg("--wager")
        if not wager_id:
            raise SystemExit("Missing --wager")
        res = call_contract(CONTRACT_ADDRESS, "get_status", [wager_id])
    else:
        raise SystemExit(f"Unknown command: {cmd}")

    print(json.dumps(res, indent=2))


if __name__ == "__main__":
    main(sys.argv)
